### PR TITLE
(NETDEV-46) Add logfile to syslog_settings

### PR DESCRIPTION
**Pull request by willmeek**
This commit adds the logfile_name, logfile_severity_level and logfile_size fields to syslog_settings for Cisco Nexus

### CODE REVIEW

**Comment by willmeek** [comment]
Customer request. Requires the release of netdev-stdlib and the merging of https://github.com/cisco/cisco-network-node-utils/pull/590

File: README.md [hunk 1]
```diff
@@ -5765,12 +5765,42 @@ Interface to send syslog data from, e.g. "management".  Valid value is a string.
 
 #### Parameters
 
-##### `name`
```
Comment (mikewiebe) [inline README.md]: Please add a caveats table and indicate that these properties are being added in version `1.10.0` of the module.

Here is an example:

https://github.com/cisco/cisco-network-puppet-module/blob/86132da1bbf4c4b34199675fb5e286100e68f0df/README.md#caveats-5

Also please add a link under the Caveats column in the support matrix.

https://github.com/cisco/cisco-network-puppet-module#resource-platform-support-matrix

File: README.md [hunk 2]
```diff
@@ -5765,12 +5765,42 @@ Interface to send syslog data from, e.g. "management".  Valid value is a string.
...
+
+Logfile severity level [0-7]
+
+##### `logfile_name`
+
+Logfile file name to use or 'unset'
+
+##### `logfile_size`
+
+Logging file maximum size [4096-2147483647]
```
Comment (mikewiebe) [inline README.md]: The range will probably be dynamic on different NX-OS platforms so I would omit the range here.

**Comment by willmeek** [comment]
@shermdog Good catch. I had assumed that the purge hash contained the logfile name, however if the size was the only change then this was not the case. Grab from the resource existing settings if that is the case. Added a beaker test.

File: README.md (hunk 1 above)
Comment (willmeek) [reply README.md]: Thanks Mike, I have added the Caveat and link

File: README.md (hunk 2 above)
Comment (willmeek) [reply README.md]: Thanks Mike, removed the range (which simplifies the validation regex hugely! :) )

**Comment by willmeek** [comment]
Travis seems to be failing on the 2.2.0 tests?

`Gem::RuntimeRequirementNotMetError: net-telnet requires Ruby version >= 2.3.0.
The current ruby version is 2.2.0.
An error occurred while installing net-telnet (0.2.0), and Bundler`

**Comment by mikewiebe** [comment]
@willmeek Have you completed the changes and are just waiting for a final review?

**Comment by willmeek** [comment]
@mikewiebe The changes should be there, awaiting final review, let me know if there is anything that needs changed / fixed, thanks! :)

File: lib/puppet/provider/syslog_settings/cisco.rb [hunk 3]
```diff
@@ -102,17 +105,61 @@ def exists?
...
          "This provider only supports a namevar of 'default'" unless @resource[:name].to_s == 'default'
-
+    fail ArgumentError,
+         'This provider requires that a logfile_name and logfile_severity_level are both specified in order '\
+         'to set logfile settings.' if @resource[:logfile_name].to_s != 'unset' &&
+                                       ((@resource[:logfile_name] && !@resource[:logfile_severity_level]) ||
+                                       (@resource[:logfile_severity_level] && !@resource[:logfile_name]))
+    fail ArgumentError,
+         'This provider requires that a logfile_name and logfile_severity_level are both specified in order '\
+         'to set logfile size.' if @resource[:logfile_size] && !@resource[:logfile_name] && !@resource[:logfile_severity_level]
```
Comment (mikewiebe) [inline lib/puppet/provider/syslog_settings/cisco.rb]: *nit* `/logfile size/logfile_size/`

**Comment by mikewiebe** [comment]
One additional nit, otherwise 👍 Will merge once Rick approves.

File: lib/puppet/provider/syslog_settings/cisco.rb (hunk 3 above)
Comment (willmeek) [reply lib/puppet/provider/syslog_settings/cisco.rb]: Thanks Mike, updated

**Comment by shermdog** [comment]
@willmeek looks like the default resource output is clashing with current netdev_stdlib type definition:

```Puppet
[root@puppet-device-devel ~]# puppet device -v -t nxapi --resource syslog_settings --trace 
Info: retrieving resource: syslog_settings from nxapi at file:///root/nxapi.yaml
syslog_settings { 'default':
  console                => 2,
  logfile_name           => 'unset',
  logfile_severity_level => 'unset',
  logfile_size           => 'unset',
  monitor                => '5',
  source_interface       => ['unset'],
  time_stamp_units       => 'seconds',
}
```

```
Error: Parameter logfile_severity_level failed on Syslog_settings[default]: Munging failed for value "unset" in class logfile_severity_level: invalid value for Integer(): "unset" (file: /root/syslog.pp, line: 1)
/etc/puppetlabs/code/environments/production/modules/netdev_stdlib/lib/puppet/type/syslog_settings.rb:80:in `Integer'
/etc/puppetlabs/code/environments/production/modules/netdev_stdlib/lib/puppet/type/syslog_settings.rb:80:in `block (3 levels) in <top (required)>'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/parameter.rb:423:in `munge'
```

It also looks like 'unset' is being configured as a long destination when logfile_name is 'unset', but severity level is set

```Puppet
syslog_settings {'default':
      console                => 'unset',
      monitor                => 'unset',
      source_interface       => 'unset',
      time_stamp_units       => 'seconds',
      logfile_name           => 'unset',
      logfile_severity_level => '3',
      logfile_size           => 'unset',
    }
```

```
Notice: /Stage[main]/Main/Syslog_settings[default]/logfile_severity_level: logfile_severity_level changed 'unset' to 3
Debug: Set state using data format 'cli'
Debug:   to value(s):
     logging logfile unset 3 
Debug: Input (cli_conf): ' logging logfile unset 3 '

nxos-local-1# sh run | inc log
logging logfile unset 3
```

**Comment by willmeek** [comment]
@shermdog Thanks for the thoroughness :)

I have re-added 'unset' to logfile_severity_level in netdev stdlib https://github.com/puppetlabs/netdev_stdlib/pull/47

And have amended the validation here to catch if we are trying to unset without unsetting the logfile_name (to avoid possible idempotency issues)

'unset' for the name not being 'unset' was the resource to property_flush field logic being in the wrong place.

**Pull request merged**